src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java

```diff
@@ -727,10 +778,17 @@ void handlePublish(MqttPublishMessage msg) {
   void handlePubrel(int pubrelMessageId) {
 
     synchronized (this.connection) {
-      this.publishComplete(pubrelMessageId);
+      MqttMessage message = qos2inbound.get(pubrelMessageId);
+
+      if (message == null) {
+        return;
```

Comment:
Ditto as above. Add a log warning here, because a PUBREL is arrived but we don't have the corresponding message-id in the qos2inbound queue ... it's a weird situation :-)